ledger/src/blockstore.rs

```diff
@@ -1921,6 +1923,121 @@ impl Blockstore {
             .map(|signatures| signatures.iter().map(|(_, signature)| *signature).collect())
     }
 
+    pub fn get_confirmed_signatures_for_address2(
+        &self,
+        address: Pubkey,
+        highest_confirmed_root: Slot,
+        before: Option<Signature>,
+        limit: usize,
+    ) -> Result<Vec<ConfirmedTransactionStatusWithSignature>> {
+        datapoint_info!(
+            "blockstore-rpc-api",
+            (
+                "method",
+                "get_confirmed_signatures_for_address2".to_string(),
+                String
+            )
+        );
+
+        // Figure the `slot` to start listing signatures at, based on the ledger location of the
+        // `before` signature if present.  Also generate a HashSet of signatures that should
+        // be excluded from the results.
+        let (mut slot, mut excluded_signatures) = match before {
+            None => (highest_confirmed_root, None),
+            Some(before) => {
+                let transaction_status = self.get_transaction_status(before)?;
+                match transaction_status {
+                    None => return Ok(vec![]),
+                    Some((slot, _)) => {
+                        let confirmed_block = self
+                            .get_confirmed_block(slot, Some(UiTransactionEncoding::Binary))
+                            .map_err(|err| {
+                                BlockstoreError::IO(IOError::new(
+                                    ErrorKind::Other,
+                                    format!("Unable to get confirmed block: {}", err),
+                                ))
+                            })?;
+
+                        // Load all signatures for the block
+                        let mut slot_signatures: Vec<_> = confirmed_block
+                            .transactions
+                            .iter()
+                            .filter_map(|transaction_with_meta| {
+                                if let Some(transaction) =
+                                    transaction_with_meta.transaction.decode()
+                                {
+                                    transaction.signatures.into_iter().next()
+                                } else {
+                                    None
+                                }
+                            })
+                            .collect();
+
+                        // Sort signatures as a way to entire a stable ordering within a slot, as
+                        // `self.find_address_signatures()` is ordered by signatures ordered and
+                        // not by block ordering
+                        slot_signatures.sort();
```

Comment:
I don't grok this sort. Doesn't it defeat the purpose?

Reply:
Within a slot, the signatures that come back from `find_address_signatures` are also sorted by signature.  So without the sort here, the `slot_signatures.truncate()` that follows will lop off the wrong signatures some of the time.

It would be so much better to remove this, and have `find_address_signatures` return in ledger order, but that's an `address_signatures_cf` format change that I'm trying to avoid here

Reply:
I'm still confused. You've just called get_confirmed_block and filter_mapped to get the signatures in ledger order ... in order, I presumed, to identify which signatures actually happened after the `before` signature and include those in the `excluded_signatures` HashSet.

If you do this sort and then truncate as written, I think you're just finding the signatures in the slot that happen to have a smaller first byte than `before` and excluding those.

I think if you just eliminate this sort line, it will work properly.